### PR TITLE
Basic FFI for casemapping

### PR DESCRIPTION
**Pull request by Manishearth**
Part of https://github.com/unicode-org/icu4x/issues/3234

Had this old PR lying around waiting for FFI to stop churning so much.

This isn't all of FFI since I need to add case-closure stuff, but that is worth doing in a separate PR.

Take care to note if I am using any outdated patterns for FFI, I *think* I updated everything but I'm not sure.

### CODE REVIEW

File: ffi/diplomat/src/casemap.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+    #[diplomat::opaque]
+    #[diplomat::rust_link(icu::casemap::CaseMapper, Struct)]
+    pub struct ICU4XCaseMapper(pub CaseMapper);
+
+    impl ICU4XCaseMapper {
+        /// Construct a new ICU4XCaseMapper instance for NFC
+        #[diplomat::rust_link(icu::casemap::ComposingNormalizer::try_new_nfc_unstable, FnInStruct)]
+        pub fn create(provider: &ICU4XDataProvider) -> Result<Box<ICU4XCaseMapper>, ICU4XError> {
+            Ok(Box::new(ICU4XCaseMapper(CaseMapper::try_new_unstable(
```
Comment (robertbastian) [inline ffi/diplomat/src/casemap.rs:23]: You need to use the`call_constructor` macro

Comment (Manishearth) [reply ffi/diplomat/src/casemap.rs:23]: blergh

File: ffi/diplomat/src/casemap.rs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    #[diplomat::opaque]
+    #[diplomat::rust_link(icu::casemap::CaseMapper, Struct)]
+    pub struct ICU4XCaseMapper(pub CaseMapper);
+
+    impl ICU4XCaseMapper {
+        /// Construct a new ICU4XCaseMapper instance for NFC
+        #[diplomat::rust_link(icu::casemap::CaseMapper::new, FnInStruct)]
+        pub fn create(provider: &ICU4XDataProvider) -> Result<Box<ICU4XCaseMapper>, ICU4XError> {
+            Ok(Box::new(ICU4XCaseMapper(call_constructor!(
+                CaseMapper::new [r => Ok(r)],
```
Comment (robertbastian) [inline ffi/diplomat/src/casemap.rs:24]: thought: should the compiled data constructor return a `CaseMapperBorrowed` type?

**Review by robertbastian — APPROVED** [review]

File: ffi/diplomat/src/casemap.rs (hunk 2 above)
Comment (Manishearth) [reply ffi/diplomat/src/casemap.rs:24]: seems unnecessary to me

**Pull request merged**